system/lib/wasmfs/backends/fetch_backend.cpp

```diff
@@ -60,38 +93,65 @@ class FetchDirectory : public MemoryDirectory {
   }
 };
 
-class FetchBackend : public ProxiedAsyncJSBackend {
-  std::string baseUrl;
+std::shared_ptr<DataFile> FetchBackend::createFile(mode_t mode) {
+  return std::make_shared<FetchFile>("", mode, this, proxy);
+}
 
-public:
-  FetchBackend(const std::string& baseUrl,
-               std::function<void(backend_t)> setupOnThread)
-    : ProxiedAsyncJSBackend(setupOnThread), baseUrl(baseUrl) {}
+std::shared_ptr<Directory> FetchBackend::createDirectory(mode_t mode) {
+  return std::make_shared<FetchDirectory>("", mode, this, proxy);
+}
 
-  std::shared_ptr<DataFile> createFile(mode_t mode) override {
-    return std::make_shared<FetchFile>(baseUrl, mode, this, proxy);
+const std::string FetchBackend::getFileURL(const std::string& filePath) {
+  if(manifest) {
+    if (auto search = manifest->find(filePath); search != manifest->end()) {
+      return baseUrl + "/" + search->second;
+    }
   }
-
-  std::shared_ptr<Directory> createDirectory(mode_t mode) override {
-    return std::make_shared<FetchDirectory>(baseUrl, mode, this, proxy);
+  if(filePath == "") {
+    return baseUrl;
   }
-};
+  return baseUrl + "/" + filePath;
+}
+uint32_t FetchBackend::getChunkSize() {
+  return chunkSize;
+}
 
 extern "C" {
-backend_t wasmfs_create_fetch_backend(const char* base_url) {
+  backend_t wasmfs_create_fetch_backend(const char* base_url, uint32_t chunkSize, FetchManifest *manifest) {
   // ProxyWorker cannot safely be synchronously spawned from the main browser
   // thread. See comment in thread_utils.h for more details.
   assert(!emscripten_is_main_browser_thread() &&
          "Cannot safely create fetch backend on main browser thread");
   return wasmFS.addBackend(std::make_unique<FetchBackend>(
     base_url ? base_url : "",
+    chunkSize != 0 ? chunkSize : DEFAULT_CHUNK_SIZE,
+    manifest,
     [](backend_t backend) { _wasmfs_create_fetch_backend_js(backend); }));
-}
+  }
 
 const char* EMSCRIPTEN_KEEPALIVE _wasmfs_fetch_get_file_path(void* ptr) {
   auto* file = reinterpret_cast<wasmfs::FetchFile*>(ptr);
   return file ? file->getPath().data() : nullptr;
 }
+const char* EMSCRIPTEN_KEEPALIVE _wasmfs_fetch_get_file_url(void* ptr) {
+  auto* file = reinterpret_cast<wasmfs::FetchFile*>(ptr);
+  return file ? file->getURL().data() : nullptr;
+}
+uint32_t EMSCRIPTEN_KEEPALIVE _wasmfs_fetch_get_chunk_size(void* ptr) {
+  auto* file = reinterpret_cast<wasmfs::FetchFile*>(ptr);
+  return file ? file->getChunkSize() : DEFAULT_CHUNK_SIZE;
+}
+
+void *EMSCRIPTEN_KEEPALIVE wasmfs_fetch_create_manifest() {
```

Comment:
Do these need `EMSCRIPTEN_KEEPALIVE` on them?

Reply:
I'm not sure, why would or wouldn't it?  I just tried to mimic other extern C functions I saw nearby.

Reply:
I removed it.